raft topology: send barrier_and_drain to a decommissioning node

patjed41 opened the pull request:
We didn't send the `barrier_and_drain` command to a
decommissioning node that could still be coordinating requests. It
could happen that a decommissioning node sent a request with an
old topology version after normal nodes received the new fence
version. Then, the request would fail on replicas with the stale
topology exception.

This PR fixes this problem by modifying `exec_global_command`.
From now on, it sends `barrier_and_drain` to a decommissioning
node.

We also stop filtering stale topology exceptions in
`test_topology_ops`. We added this filter after detecting the bug
fixed by this PR.

Fixes scylladb/scylladb#15804
Fixes scylladb/scylladb#16579
Fixes scylladb/scylladb#16642

patjed41 commented:
#16579 is about two errors -- one is fixed by this PR and the other one has already been fixed. So, I'm assuming this PR can mark this issue as fixed.

We ignore error messages containing "stale topology exception" in dtests. Once this PR is merged, I'll revert it.

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -2620,6 +2620,13 @@ future<service::topology> system_keyspace::load_topology_state() {
             map = &ret.new_nodes;
         } else {
             map = &ret.transition_nodes;
+            // At most one node at a time can be in transitioning state.
```
The point of it being a map is that if we will want, at some point, to support more then one operation at  a time we will not have to change all of the code to map since it will already assume that it is a map. By dropping the generic code and insert the assumptions that there is only one element in the map you removed all the benefit and left all the disadvantages.  I do not see the point in most of this patch. May be only the first check to make sure that there is only one node is there for now.

patjed41 replied:
OK, I will leave only the first check. It won't be needed after doing the changes from the comment below, but I think it's still a good idea because if we currently have more than one transitioning node, then something must have gone wrong.

I would also like to extend the comment above `transition_nodes` to something like "Currently at most one node at a time will be here, but the code shouldn't assume it because we might support more than one operation at a time in the future". It's not clear why this map is here and how it should be handled.

gleb-cloudius replied:
> OK, I will leave only the first check. It won't be needed after doing the changes from the comment below, but I think it's still a good idea because if we currently have more than one transitioning node, then something must have gone wrong.

Yes, that what I wrote as well.

> 
> I would also like to extend the comment above `transition_nodes` to something like "Currently at most one node at a time will be here, but the code shouldn't assume it because we might support more than one operation at a time in the future". It's not clear why this map is here and how it should be handled.

OK.

kbr-scylla replied:
> The point of it being a map is that if we will want, at some point, to support more then one operation at a time

To be honest I don't think we will. Token ring changes? Will be quick and trivial once all data moves to tablets (we can skip streaming/repair altogether), so no need to parallelize there.

kbr-scylla replied:
Assuming a single transitioning node simplifies reasoning and code.

gleb-cloudius replied:
> > The point of it being a map is that if we will want, at some point, to support more then one operation at a time
> 
> To be honest I don't think we will. Token ring changes? Will be quick and trivial once all data moves to tablets (we can skip streaming/repair altogether), so no need to parallelize there.

I also do not think (hope) that we will ever support parallel token ring changes, but we may be forced to support running in parallel several non token ring changing operations like rebuild.

patjed41 replied:
Since keeping the map might be useful, I did as I proposed earlier -- I left only the sanity check and extended the comment.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5804/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7169/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5804/)
#### Build Details:
- Duration: 2 hr 23 min
- Builder: i-0f409a122d5c8d11c (m5d.12xlarge)

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1453,11 +1453,22 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+        // We must send barrier_and_drain to the decommissioning node as it might be coordinating requests.
+        auto it = _topo_sm._topology.transition_nodes.begin();
+        auto node_is_decommissioning = it != _topo_sm._topology.transition_nodes.end()
+                && (it->second.state == node_state::decommissioning || it->second.state == node_state::left_token_ring);
```
The problem is that failed bootstrap and replace also go through the `node_state::left_token_ring`. Not sure we want to send the barrier to then though especially as they may be dead by now.

patjed41 replied:
We probably shouldn't send the barrier to them. I'll investigate what exactly happens and think about a fix if needed.

patjed41 replied:
The following happens after we start the rollback of a bootstrap or replace and the joining node is dead:
- the `left_token_ring` state handler in `handle_node_transition` calls `global_token_metadata_barrier`,
- `global_token_metadata_barrier` sends `barrier_and_drain`, which fails because it includes the dead joining node,
- `global_token_metadata_barrier` catches the exception, proceeds and sends `barrier`, which behaves the same way before and after the change because the joining node is not included.

So, IIUC, the `left_token_ring` state handler for a failed bootstrap or decommission should succeed after the change if it would succeed before the change. We will log an error after failed `barrier_and_drain`, but we are on the error path, so some errors have already been logged. In conclusion, I don't think this change will have a noticeable bad impact.

To fix this, we need a way to distinguish a finishing decommission from a failed bootstrap or replace. The only idea I have is introducing a new state for a failed bootstrap and replace.

I opened #16871. This is a small issue, but the bottom part of if -- "Another issue is in `handle_node_transition`" -- is another reason why we need a way to distinguish these two cases.

I can introduce this state in this PR or as a follow-up. It should be easy and add only a few lines of code. I can also implement a different solution if someone has an idea. cc @kbr-scylla

For now, I'll do fixes from the other comments and send a new version.

kbr-scylla replied:
I have nothing against introducing a new state. And if we want to introduce a new state we should do it before branching. @gleb-cloudius please give your opinion.

gleb-cloudius replied:
> The only idea I have is introducing a new state for a failed bootstrap and replace.

Another is to make `left_token_ring` to be a transition state, not a node state. @kbr-scylla do you remember what is the reason it is not?

kbr-scylla replied:
```
commit b8ddfd9ef977d074fbc49fc8becac6e651b95a81
Author: Kamil Braun <kbraun@scylladb.com>
Date:   Fri May 19 17:30:43 2023 +0200

    raft topology: introduce `left_token_ring` state
    
    We want for the decommissioning node to wait before shutting down until
    every node learns that it left the token ring. Otherwise some nodes may
    still try coordinating writes to that nodes after it already shut down,
    leading to unnecessary failures on the data path(e.g. for CL=ALL writes).
    
    Before this change, a node would shut down immediately after observing
    that it was in `left` state; some other nodes may still see it in
    `decommissioning` state and the topology transition state as
    `write_both_read_new`, so they'd try to write to that node.
    
    After this change, the node first enters the `left_token_ring` state
    before entering `left`, while the topology transition state is removed
    (so we've finished the token ring change - the node no longer has tokens
    in the ring, but it's still part of the topology). There we perform a
    read barrier, allowing all nodes to observe that the decommissioning
    node has indeed left the token ring. Only after that barrier succeeds we
    allow the node to shut down.

```

The purpose of left_token_ring was to remove the node's tokens from the token ring before having it shut down. I guess it felt more natural to use node_state than transition_state to decide whether a node should keep its tokens. Perhaps combining `node_state == decommissioning && transition_state == left_token_ring` would achieve the same thing. And then we could execute different logic for `node_state == bootstrapping && transition_state == left_token_ring` (this would happen during rollback), etc.

gleb-cloudius replied:
> And then we could execute different logic for node_state == bootstrapping && transition_state == left_token_ring (this would happen during rollback), etc.

Yes. Today we lose info about what operation is ongoing when we move to `left_token_ring`. Introducing yet another `left_token_ring_after_rollback` is a way to preserve it, but making `left_token_ring` a transition state feels more correct.

patjed41 replied:
Then I'll make `left_token_ring` a transition state. Should I do it in this PR or can we leave it as an immediate follow-up? Leaving this as a follow-up would allow us to stop ignoring stale topology exceptions in tests earlier.

kbr-scylla replied:
IMO follow-up

gleb-cloudius replied:
> IMO follow-up

OK

patjed41 commented on service/storage_service.cc:
```diff
@@ -1453,11 +1453,22 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
-            | boost::adaptors::filtered([&exclude_nodes] (const std::pair<const raft::server_id, replica_state>& n) {
-                return !exclude_nodes.contains(n.first);
-            })
-            | boost::adaptors::map_keys;
```
I'll rewrite it using `boost::range::join` and without the assumption that there is only one transitioning node.

gleb-cloudius replied:
Thank. Not that I care about the assumption (as long as it is documented), but IMO without the assumption the code will be even simpler.

patjed41 commented:
v2:
- `transition_nodes` -- the first commit:
  - removed the code changes that assume there is only one transitioning node at a time
  - extended the comment describing `transition_nodes`
  - left the new sanity check
- `exec_global_command` -- the second commit:
  - refactored in a way that doesn't make the code assume there is only one transitioning node
  - the transitioning nodes can also be excluded now; although we don't need it now, it might be useful in the future and the implementation is simpler this way

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5919/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7260/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5919/)
#### Failed Tests (2/23802):
- [test_tablet_missing_data_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5919/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_missing_data_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_missing_data_repair)
- [topology_experimental_raft.test_tablets.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5919/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.5)
#### Build Details:
- Duration: 1 hr 58 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented:
Flaky test:
https://github.com/scylladb/scylladb/issues/16859

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5925/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5925/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7264/)
#### Failed Tests (2/23803):
- [test_scrub_with_multi_nodes_expect_data_rebuild](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5925/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_scrub_with_multi_nodes_expect_data_rebuild) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scrub_with_multi_nodes_expect_data_rebuild)
- [test_scrub_with_multi_nodes_expect_data_rebuild](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5925/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_scrub_with_multi_nodes_expect_data_rebuild_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scrub_with_multi_nodes_expect_data_rebuild)
#### Build Details:
- Duration: 2 hr 34 min
- Builder: spider6.cloudius-systems.com

patjed41 commented:
Rebased to fix a conflict

patjed41 commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5925/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5925/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7264/)
> #### Failed Tests (2/23803):
> 
>     * [test_scrub_with_multi_nodes_expect_data_rebuild](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5925/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_scrub_with_multi_nodes_expect_data_rebuild) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scrub_with_multi_nodes_expect_data_rebuild)
> 
>     * [test_scrub_with_multi_nodes_expect_data_rebuild](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5925/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_scrub_with_multi_nodes_expect_data_rebuild_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scrub_with_multi_nodes_expect_data_rebuild)
> 
> 
> #### Build Details:
> 
>     * Duration: 2 hr 34 min
> 
>     * Builder: spider6.cloudius-systems.com

Reopened #15195

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5936/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7275/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5936/)
#### Failed Tests (4/23802):
- [test_partial_upgrade_can_be_finished_with_removenode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5936/testReport/junit/%28root%29/test_raft_cluster_features/Tests___Unit_Tests___test_partial_upgrade_can_be_finished_with_removenode) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_partial_upgrade_can_be_finished_with_removenode)
- [test_recover_stuck_raft_recovery](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5936/testReport/junit/%28root%29/test_raft_recovery_stuck/Tests___Unit_Tests___test_recover_stuck_raft_recovery) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_recover_stuck_raft_recovery)
- [topology_experimental_raft.test_raft_cluster_features.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5936/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_raft_cluster_features_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_raft_cluster_features.debug.1)
- [topology_custom.test_raft_recovery_stuck.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5936/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_raft_recovery_stuck_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_recovery_stuck.dev.3)
#### Build Details:
- Duration: 1 hr 57 min
- Builder: spider1.cloudius-systems.com

kostja commented:
It's unfortunate we can't come up with a functional test for the problem.

kostja commented:
Perhaps a dedicated dtest that would create a cluster of 3 nodes, run decommission, and observe that no query failed while the node is decommissioned? The test would have to set the query timeout to infinity though, to make sure no query fails do to a timeout.

patjed41 commented:
> Perhaps a dedicated dtest that would create a cluster of 3 nodes, run decommission, and observe that no query failed while the node is decommissioned?

`test_topology_ops.py` does that, but it runs all topology operations, not only decommission. After uncommenting the part disabled due to the bug this PR fixes, this test fails very often, even locally.

kbr-scylla commented:
So we already have a functional test, it's enough.

patjed41 commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5936/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7275/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5936/)
> #### Failed Tests (4/23802):
> 
>     * [test_partial_upgrade_can_be_finished_with_removenode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5936/testReport/junit/%28root%29/test_raft_cluster_features/Tests___Unit_Tests___test_partial_upgrade_can_be_finished_with_removenode) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_partial_upgrade_can_be_finished_with_removenode)
> 
>     * [test_recover_stuck_raft_recovery](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5936/testReport/junit/%28root%29/test_raft_recovery_stuck/Tests___Unit_Tests___test_recover_stuck_raft_recovery) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_recover_stuck_raft_recovery)
> 
>     * [topology_experimental_raft.test_raft_cluster_features.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5936/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_raft_cluster_features_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_raft_cluster_features.debug.1)
> 
>     * [topology_custom.test_raft_recovery_stuck.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5936/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_raft_recovery_stuck_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_recovery_stuck.dev.3)
> 
> 
> #### Build Details:
> 
>     * Duration: 1 hr 57 min
> 
>     * Builder: spider1.cloudius-systems.com

The `test_raft_recovery_stuck` failure looks like one of the "driver dropping connections" failures:
```
E           cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.251.110.14:9042 datacenter1>: ConnectionException('Host has been marked down or removed')})
```
We have multiple issue open with the message above.

I can't explain the `test_raft_cluster_features` failure for now. It's a timeout in `wait_for_cql` after restarting the cluster. Maybe it's a reappearance of #14679, which was closed after the failure stopped appearing.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5975/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5975/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7309/)
#### Failed Tests (2/23829):
- [test_snapshot](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5975/testReport/junit/%28root%29/test_snapshot/Tests___Unit_Tests___test_snapshot) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_snapshot)
- [topology.test_snapshot.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5975/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_snapshot_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_snapshot.debug.1)
#### Build Details:
- Duration: 2 hr 37 min
- Builder: i-0f3d962665b0a32cf (m5d.12xlarge)

patjed41 commented:
The last CI failure is #16926. It's a different test, but the error happens during bootstrap of initial nodes.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6001/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7331/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6001/)
#### Build Details:
- Duration: 1 hr 52 min
- Builder: spider3.cloudius-systems.com

patjed41 commented:
Finally CI passed

@gleb-cloudius Could you review it?

gleb-cloudius reviewed (approved):
So I understand that this still sends barrier to a bootstrapping node during rollback, but the claim is that this does not cause any serious issues except more errors printed in the logs and it will be addressed in the followup.

patjed41 commented:
> So I understand that this still sends barrier to a bootstrapping node during rollback, but the claim is that this does not cause any serious issues except more errors printed in the logs and it will be addressed in the followup.

Yes, I'll open a follow-up issue.

kbr-scylla commented:
@patjed41 please rebase, topology coordinator was moved to separate file

patjed41 commented:
> @patjed41 please rebase, topology coordinator was moved to separate file

Done

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6082/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7403/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6082/)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: spider4.cloudius-systems.com
